src/commands/lock.rs

```diff
@@ -0,0 +1,144 @@
+        Ok(results) => results,
+        Err(errors) => {
+            // SAFETY: this is a non-empty, so it must always have a first
+            // element.
+            bail!(errors.into_iter().next().unwrap())
+        }
+    };
+
+    let mut images: HashSet<String> = HashSet::new();
+    for result in results {
+        let doc = result.document().root();
+
+        for task in doc.ast().as_v1().expect("should be a v1 document").tasks() {
+            let task_name_token = task.name();
+            let task_name = task_name_token.inner().text();
+            let doc_path = result.document().path();
+            let Some(runtime) = task.runtime() else {
+                tracing::warn!(
+                    "Skipping task {task_name} in document {doc_path} with no runtime section",
+                );
+                continue;
+            };
+            let Some(image) = runtime.container().and_then(|c| c.value().ok()) else {
+                tracing::warn!(
+                    "Skipping task {task_name} in document {doc_path} with no container image",
+                );
+                continue;
+            };
+            let Expr::Literal(LiteralExpr::String(s)) = image.expr() else {
+                tracing::warn!(
+                    "Skipping image with non-literal value in task {task_name} in document \
+                     {doc_path}",
+                );
+                continue;
+            };
+            let Some(text) = s.text() else {
+                tracing::warn!(
+                    "Skipping image with placeholder value in task {task_name} in document \
+                     {doc_path}",
+                );
+                continue;
+            };
+            let mut buffer = String::new();
+            text.unescape_to(&mut buffer);
+            images.insert(buffer);
+        }
+    }
+
+    let time = Utc::now();
+
+    let mut map: HashMap<String, String> = HashMap::new();
+    let docker = crankshaft_docker::with_defaults()?;
+
+    for image in images {
+        let prefix = image.split(':').next().unwrap_or("");
+
+        let i = docker
+            .inner()
+            .inspect_registry_image(&image, None)
+            .await
+            .expect("should inspect registry image");
+
+        // Insert the manifest digest into the map.
+        map.insert(
+            image.clone(),
+            prefix.to_owned() + "@" + &i.descriptor.digest.expect("should have a digest"),
+        );
+    }
+
+    if !map.is_empty() {
+        let lock = Lock {
+            timestamp: time.to_string(),
+            images: map,
+        };
+        let data = toml::to_string_pretty(&lock)?;
+        std::fs::write(output_path, data)?;
+    } else {
+        let lock = Lock {
+            timestamp: time.to_string(),
+            images: HashMap::new(),
+        };
+
+        let data = toml::to_string_pretty(&lock)?;
+        std::fs::write(output_path, data)?;
```

Comment:
this writes to the file twice. The if-else statement can be removed. We can just always write out the map regardless of emptiness